File: receiver-java/src/main/java/gt/edu/uvg/redes/receiver/Crc32Codec.java
package gt.edu.uvg.redes.receiver;

public final class Crc32Codec {
    private static final int CRC_BITS = 32;
    private static final int MINIMUM_DATA_BITS = 32;
    private static final long INITIAL_VALUE = 0xFFFFFFFFL;
    private static final long POLYNOMIAL = 0xEDB88320L;

    private Crc32Codec() {
    }

    public record Result(String dataBits, String status, int errorsDetected, int errorsCorrected) {
    }

    public static Result decode(String frameBits, int originalBitLength) {
        if (originalBitLength < 0 || originalBitLength % 8 != 0) {
            throw new IllegalArgumentException("La longitud ASCII debe ser múltiplo de 8");
        }
        int protectedDataLength = protectedDataLength(originalBitLength);
        if (frameBits.length() != protectedDataLength + CRC_BITS) {
            throw new IllegalArgumentException("Longitud de CRC-32 inválida");
        }
        String protectedDataBits = frameBits.substring(0, protectedDataLength);
        long expected = Long.parseUnsignedLong(frameBits.substring(protectedDataLength), 2);
        long actual = checksum(bitsToBytes(protectedDataBits));
        if (actual != expected) {
            throw new IllegalArgumentException("CRC-32 no coincide");
        }
        return new Result(protectedDataBits.substring(0, originalBitLength), "OK", 0, 0);
    }

    public static String encode(String dataBits) {
        if (dataBits.length() % 8 != 0 || !dataBits.matches("[01]*")) {
            throw new IllegalArgumentException("Los datos CRC deben ser bytes ASCII");
        }
        String protectedDataBits = padData(dataBits);
        long value = checksum(bitsToBytes(protectedDataBits));
        return protectedDataBits + String.format("%32s", Long.toBinaryString(value)).replace(' ', '0');
    }

    public static long checksum(byte[] bytes) {
        long crc = INITIAL_VALUE;
        for (byte value : bytes) {
            crc ^= value & 0xFFL;
            for (int bit = 0; bit < 8; bit++) {
                if ((crc & 1L) != 0) {
                    crc = (crc >>> 1) ^ POLYNOMIAL;
                } else {
                    crc >>>= 1;
                }
            }
        }
        return (crc ^ INITIAL_VALUE) & 0xFFFFFFFFL;
    }

    public static int protectedDataLength(int originalBitLength) {
        if (originalBitLength < 0 || originalBitLength % 8 != 0) {
            throw new IllegalArgumentException("La longitud ASCII debe ser múltiplo de 8");
        }
        return Math.max(originalBitLength, MINIMUM_DATA_BITS);
    }

    private static String padData(String dataBits) {
        int protectedLength = protectedDataLength(dataBits.length());
        return dataBits + "0".repeat(protectedLength - dataBits.length());
    }

    private static byte[] bitsToBytes(String bits) {
        byte[] bytes = new byte[bits.length() / 8];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = (byte) Integer.parseInt(bits.substring(i * 8, i * 8 + 8), 2);
        }
        return bytes;
    }
}

File: sender-python/sender/crc32_codec.py
import zlib
from dataclasses import dataclass

_CRC_BITS = 32
_MINIMUM_DATA_BITS = 32


@dataclass(frozen=True)
class Result:
    data_bits: str
    status: str
    errors_detected: int
    errors_corrected: int


def protected_data_length(original_bit_length: int) -> int:
    if original_bit_length < 0 or original_bit_length % 8 != 0:
        raise ValueError("La longitud ASCII debe ser múltiplo de 8")
    return max(original_bit_length, _MINIMUM_DATA_BITS)


def checksum(data: bytes) -> int:
    return zlib.crc32(data) & 0xFFFFFFFF


def encode(data_bits: str) -> str:
    if len(data_bits) % 8 != 0 or not set(data_bits) <= {"0", "1"}:
        raise ValueError("Los datos CRC deben ser bytes ASCII")
    protected_data_bits = _pad_data(data_bits)
    value = checksum(_bits_to_bytes(protected_data_bits))
    return protected_data_bits + format(value, "032b")


def decode(frame_bits: str, original_bit_length: int) -> Result:
    protected_length = protected_data_length(original_bit_length)
    if len(frame_bits) != protected_length + _CRC_BITS:
        raise ValueError("Longitud de CRC-32 inválida")
    protected_data_bits = frame_bits[:protected_length]
    expected = int(frame_bits[protected_length:], 2)
    actual = checksum(_bits_to_bytes(protected_data_bits))
    if actual != expected:
        raise ValueError("CRC-32 no coincide")
    return Result(protected_data_bits[:original_bit_length], "OK", 0, 0)


def _pad_data(data_bits: str) -> str:
    protected_length = protected_data_length(len(data_bits))
    return data_bits + "0" * (protected_length - len(data_bits))


def _bits_to_bytes(bits: str) -> bytes:
    return bytes(int(bits[i:i + 8], 2) for i in range(0, len(bits), 8))
